### PR TITLE
Noise synthesizer

### PR DESCRIPTION
**Pull request by dissonancedev**

### CODE REVIEW

**Comment by djowel** [comment]
Oh, the implementation is rather expensive, by the looks of it. There should be a better (more performant) way to do this for real-time. Where did you get the algorithm? Why is there a need for cos, log and sqrt? Isn't white-noise just random values (with equal distribution) from -1.0 to 1.0? As in https://www.dsprelated.com/freebooks/sasp/White_Noise.html ? 

Here's a fast white noise generator (using a very fast pseudo random number generator with xor -- enough to fool the ears. Keep in mind that we don't use it for things like crypto, so what's good enough can be significantly relaxed): 
https://www.musicdsp.org/en/latest/Synthesis/216-fast-whitenoise-generator.html

As noted in the comment:
"To get pink noise, you can apply a 3dB/Oct filter, for example the pink noise filter in the Filters section.
To get brown noise, apply an one pole LP filter to get a 6dB/oct slope.
Peter"

**Comment by dissonancedev** [comment]
Thank you for the feedback. I have updated wthe code with your suggestions. They sound better now and should also perform significantly better too.

File: q_lib/include/q/synth/noise.hpp [hunk 1]
```diff
@@ -6,29 +6,30 @@
...
-      
-         return fabs(s) > 1 ? (s / fabs(s)) : s;
+         g_x1 ^= g_x2;
+         s = g_x2 * g_fScale;
+         g_x2 += g_x1;
       }
+
+      int g_x1 = 0x67452301;
+      int g_x2 = 0xefcdab89;
+      float g_fScale = 2.0f / 0xffffffff;
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp]: Make this const_expr

Comment (djowel) [reply q_lib/include/q/synth/noise.hpp]: Oh and please don't use camelCase. use_underscores_instead.

File: q_lib/include/q/synth/noise.hpp [hunk 2]
```diff
@@ -40,59 +41,58 @@ namespace cycfi::q
...
          for (unsigned int i = 0; i < 7; ++i) {
-            b[0] = k[0] * white + k[0] * b[0];
+            b[i] = k[i] * (white + b[i]);
          }
-         
-         float pink = (b[0] + b[1] + b[2] + b[3] + b[4] + b[5] + white - b[6]);
-         
-         return fabs(pink) > 1 ? (pink / fabs(pink)) : pink;
+
+         pink = 0.05f * (b[0] + b[1] + b[2] + b[3] + b[4] + b[5] + white - b[6]);
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp:63]: Can this be optimized using IIR? Is this how Music DSP implementation does it?
(update, now I see the answer is yes)

File: q_lib/include/q/synth/noise.hpp [hunk 3]
```diff
@@ -6,29 +6,30 @@
...
-      float operator()() const
+      float operator()()
       {
-         float r1 = (float)fast_rand() / 32767,
-            r2 = (float)fast_rand() / 32767;
-         
-         float s = (float)0.151803f * sqrt(-2.0f * fastlog(r1)) * fastcos(2.0f * M_PI * r2);
-      
-         return fabs(s) > 1 ? (s / fabs(s)) : s;
+         g_x1 ^= g_x2;
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp]: Don't forget proper attribution (citing the source)

**Review by djowel — CHANGES_REQUESTED** [review]
Some comments...

File: q_lib/include/q/synth/noise.hpp (hunk 2 above)
Comment (djowel) [reply q_lib/include/q/synth/noise.hpp:63]: Ah, that's already how they implement 3dB/oct? Interesting. I haven't done this before.

File: q_lib/include/q/synth/noise.hpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+         for (unsigned int i = 0; i < 7; ++i) {
+            b[i] = k[i] * (white + b[i]);
+         }
+
+         pink = 0.05f * (b[0] + b[1] + b[2] + b[3] + b[4] + b[5] + white - b[6]);
+
+         return pink;
+      }
+
+      float f[7] = {
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp]: Make this const_expr

**Review by djowel — CHANGES_REQUESTED** [review]

File: q_lib/include/q/synth/noise.hpp (hunk 2 above)
Comment (dissonancedev) [reply q_lib/include/q/synth/noise.hpp:63]: This is the implementation I found:
https://www.musicdsp.org/en/latest/Filters/76-pink-noise-filter.html

It is the same one that is suggested in Peter's comment that you pasted. I tried some one pole HP but they didn't sound good.

File: q_lib/include/q/synth/noise.hpp (hunk 1 above)
Comment (dissonancedev) [reply q_lib/include/q/synth/noise.hpp]: Compiler complains about this one:
` error: non-static data member ‘scale’ declared ‘constexpr’
       constexpr float scale = 2.0f / 0xffffffff;`

File: q_lib/include/q/synth/noise.hpp (hunk 4 above)
Comment (dissonancedev) [reply q_lib/include/q/synth/noise.hpp]: Same complaint here too
`error: non-static data member ‘f’ declared ‘constexpr’
       };
`

File: q_lib/include/q/synth/noise.hpp (hunk 1 above)
Comment (dissonancedev) [reply q_lib/include/q/synth/noise.hpp]: never mind I forgot to declare it static :facepalm:

File: q_lib/include/q/synth/noise.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*=============================================================================
+   Copyright (c) 2014-2021 Joel de Guzman. All rights reserved.
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp]: Feel free to add your name in there. You wrote it :-)
Just don't forget to attribute the code you got this from (e.g. comments before the class).

File: q_lib/include/q/synth/noise.hpp (hunk 4 above)
Comment (djowel) [reply q_lib/include/q/synth/noise.hpp]: make it static. constexpr has to be static.

**Comment by dissonancedev** [comment]
Alright, should be done now :)

File: q_lib/include/q/synth/noise.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+#include <q/detail/fast_math.hpp>
+#include <q/fx/lowpass.hpp>
+
+namespace cycfi::q
+{
+   ////////////////////////////////////////////////////////////////////////////
+   // white_noise_synth: Synthesizes white noise.
+   // Taken from musicdsp.org
+   // Author or source: ed.bew@hcrikdlef.dreg
```
Comment (djowel) [inline q_lib/include/q/synth/noise.hpp]: It's probably best to provide a synopsis and link to some web page here.

**Comment by djowel** [comment]
perfect!

**Pull request merged**